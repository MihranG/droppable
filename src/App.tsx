import React, {Component} from 'react';
import {Box} from "./Box";
import {css} from "@emotion/core";
import styled from "@emotion/styled";
import {appComponentsState} from "./types";

const RECTANGLE_SIZE_HARDCODED = 4 * 4;


export class App extends Component <{}, appComponentsState> {

    constructor(props: Readonly<{}>) {
        super(props);
        this.state = {
            items: Array.apply(null, Array(RECTANGLE_SIZE_HARDCODED)).map((_, i) => ({value: i + 1, isVisible: true})),
            draggingID: -1
        };
    }

    onDrop = (id: number): () => void => {
        const {items, draggingID} = this.state;

        return () => {
            const newItems = [...items];
            const buffer = newItems[id];
            newItems[id] = newItems[draggingID];
            newItems[draggingID] = buffer;

            this.setState({
                items: newItems,
                draggingID: -1
            })
        }
    };

    onDragStop = (): () => void => {
        return () => {
            this.setState({
                draggingID: -1
            })

        }
    };

    onDragStart = (id: number): () => void => () => {
        this.setState({
            draggingID: id
        })

    };

    render() {
        const {items} = this.state;

        return (
            <div css={wrapperDivStyle}>
                <StyledUl>
                    {items.map((item, index) => (

                        <Box key={`item-${item.value}`}
                             onDragStart={this.onDragStart}
                             onDragStop={this.onDragStop}
                             onDrop={this.onDrop}
                             index={index} initialItem={item}
                        />

                    ))}
                </StyledUl>

            </div>
        );
    }
}


const wrapperDivStyle = css`
    #margin: 0 50px;
`;

const StyledUl = styled.ul`
    display: flex;
    justify-content: center;
    flex-wrap: wrap;
    list-style:none;
    padding: 0;
    &.draggingClass{
        z-index:1;
    }
`;

export default App;