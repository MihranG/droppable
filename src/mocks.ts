export const initialStateMock =  {
    "draggingID": -1,
    "items":  [
        {
            "isVisible": true,
            "value": 1,
        },
        {
            "isVisible": true,
            "value": 2,
        },
        {
            "isVisible": true,
            "value": 3,
        },
        {
            "isVisible": true,
            "value": 4,
        },
        {
            "isVisible": true,
            "value": 5,
        },
        {
            "isVisible": true,
            "value": 6,
        },
        {
            "isVisible": true,
            "value": 7,
        },
        {
            "isVisible": true,
            "value": 8,
        },
        {
            "isVisible": true,
            "value": 9,
        },
        {
            "isVisible": true,
            "value": 10,
        },
        {
            "isVisible": true,
            "value": 11,
        },
        {
            "isVisible": true,
            "value": 12,
        },
        {
            "isVisible": true,
            "value": 13,
        },
        {
            "isVisible": true,
            "value": 14,
        },
        {
            "isVisible": true,
            "value": 15,
        },
        {
            "isVisible": true,
            "value": 16,
        },
    ],
};


export const boxItemPropMock = {
    isVisible: true,
    value: 5
}