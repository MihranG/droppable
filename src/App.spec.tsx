import React from 'react';
import ReactDOM from 'react-dom';
import {App} from './App';
import {create, act, /*ReactTestRenderer*/} from 'react-test-renderer';
import {initialStateMock} from "./mocks";

describe('App root component', () => {
    let AppComponent: any;

    jest.spyOn(Math,'random').mockImplementation(()=>0.5);

    it('should render without errors', () => {
        const div = document.createElement('div');
        ReactDOM.render(<App/>, div);
        ReactDOM.unmountComponentAtNode(div);
    });

    act(() => {
        AppComponent = create(<App/>);
    });

    it('should match snapshots ', () => {
        expect(AppComponent.toJSON()).toMatchSnapshot();
    });

    it('should initiate correct state', () => {
        AppComponent = create(<App/>);
        expect(AppComponent.getInstance()!.state).toEqual(initialStateMock)
    });

    describe('should has method',()=>{
        it('onDragStart which must return function',()=>{
            expect(AppComponent.getInstance()!.state.draggingID).toBe(-1);
            AppComponent.getInstance().onDragStart(8)();
            expect(AppComponent.getInstance()!.state.draggingID).toBe(8);
        });


        it('onDragStop which must return function which sets draggingId to initial value',()=>{
            AppComponent.getInstance().setState({draggingID:5});
            expect(AppComponent.getInstance()!.state.draggingID).toBe(5);
            AppComponent.getInstance().onDragStop()();
            expect(AppComponent.getInstance()!.state.draggingID).toBe(-1);
        });

        it('onDrop which must return function which will swap items',()=>{
            AppComponent.getInstance().setState({draggingID:5});
            expect(AppComponent.getInstance()!.state.draggingID).toBe(5);
            expect(AppComponent.getInstance()!.state.items[5]).toEqual({isVisible:true, value:6});
            expect(AppComponent.getInstance()!.state.items[10]).toEqual({isVisible:true, value:11});
            AppComponent.getInstance().onDrop(10)();
            expect(AppComponent.getInstance()!.state.items[10]).toEqual({isVisible:true, value:6});
            expect(AppComponent.getInstance()!.state.items[5]).toEqual({isVisible:true, value:11});
            expect(AppComponent.getInstance()!.state.draggingID).toBe(-1);
        })
    })
});
