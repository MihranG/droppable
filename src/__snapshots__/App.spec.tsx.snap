// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App root component should match snapshots  1`] = `
<div
  css={
    Object {
      "map": undefined,
      "name": "v0o5o4",
      "next": undefined,
      "styles": "
    #margin: 0 50px;
",
    }
  }
>
  <ul
    className="css-qye5tu"
  >
    <li
      className="css-12x2kaz"
      data-testid="box-1"
      draggable={true}
      onDragEnd={[Function]}
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
    >
      <span>
         
        1
      </span>
    </li>
    <li
      className="css-12x2kaz"
      data-testid="box-2"
      draggable={true}
      onDragEnd={[Function]}
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
    >
      <span>
         
        2
      </span>
    </li>
    <li
      className="css-12x2kaz"
      data-testid="box-3"
      draggable={true}
      onDragEnd={[Function]}
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
    >
      <span>
         
        3
      </span>
    </li>
    <li
      className="css-12x2kaz"
      data-testid="box-4"
      draggable={true}
      onDragEnd={[Function]}
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
    >
      <span>
         
        4
      </span>
    </li>
    <li
      className="css-12x2kaz"
      data-testid="box-5"
      draggable={true}
      onDragEnd={[Function]}
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
    >
      <span>
         
        5
      </span>
    </li>
    <li
      className="css-12x2kaz"
      data-testid="box-6"
      draggable={true}
      onDragEnd={[Function]}
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
    >
      <span>
         
        6
      </span>
    </li>
    <li
      className="css-12x2kaz"
      data-testid="box-7"
      draggable={true}
      onDragEnd={[Function]}
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
    >
      <span>
         
        7
      </span>
    </li>
    <li
      className="css-12x2kaz"
      data-testid="box-8"
      draggable={true}
      onDragEnd={[Function]}
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
    >
      <span>
         
        8
      </span>
    </li>
    <li
      className="css-12x2kaz"
      data-testid="box-9"
      draggable={true}
      onDragEnd={[Function]}
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
    >
      <span>
         
        9
      </span>
    </li>
    <li
      className="css-12x2kaz"
      data-testid="box-10"
      draggable={true}
      onDragEnd={[Function]}
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
    >
      <span>
         
        10
      </span>
    </li>
    <li
      className="css-12x2kaz"
      data-testid="box-11"
      draggable={true}
      onDragEnd={[Function]}
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
    >
      <span>
         
        11
      </span>
    </li>
    <li
      className="css-12x2kaz"
      data-testid="box-12"
      draggable={true}
      onDragEnd={[Function]}
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
    >
      <span>
         
        12
      </span>
    </li>
    <li
      className="css-12x2kaz"
      data-testid="box-13"
      draggable={true}
      onDragEnd={[Function]}
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
    >
      <span>
         
        13
      </span>
    </li>
    <li
      className="css-12x2kaz"
      data-testid="box-14"
      draggable={true}
      onDragEnd={[Function]}
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
    >
      <span>
         
        14
      </span>
    </li>
    <li
      className="css-12x2kaz"
      data-testid="box-15"
      draggable={true}
      onDragEnd={[Function]}
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
    >
      <span>
         
        15
      </span>
    </li>
    <li
      className="css-12x2kaz"
      data-testid="box-16"
      draggable={true}
      onDragEnd={[Function]}
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
    >
      <span>
         
        16
      </span>
    </li>
  </ul>
</div>
`;
